### PR TITLE
refactor!: observability API and OTeL provider

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

n/a

## Description of changes

This is the final PR to main that refactors the observability instrumentation (telemetry) APIs. This also includes an implementation of a `TelemetryProvider` based on OpenTelemetry.

Incremental PRs already reviewed:

* https://github.com/awslabs/smithy-kotlin/pull/863
* https://github.com/awslabs/smithy-kotlin/pull/864
* https://github.com/awslabs/smithy-kotlin/pull/866
* https://github.com/awslabs/smithy-kotlin/pull/869
* https://github.com/awslabs/smithy-kotlin/pull/871
* https://github.com/awslabs/smithy-kotlin/pull/872
* https://github.com/awslabs/smithy-kotlin/pull/874
* https://github.com/awslabs/smithy-kotlin/pull/881
* https://github.com/awslabs/smithy-kotlin/pull/885

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by aajtodd** [comment]
TODO: Remove or update existing tracing/observability design.

**Comment by ianbotsf** [comment]
> TODO: Remove or update existing tracing/observability design.

I recommend to update, which I believe may help users write their own observability integrations and/or use observability hooks.

File: .changes/44708ceb-cc8c-478c-bd35-af82aa79f546.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "id": "44708ceb-cc8c-478c-bd35-af82aa79f546",
+    "type": "misc",
+    "description": "**BREAKING**: Refactor observability API and configuration."
```
Comment (ianbotsf) [inline .changes/44708ceb-cc8c-478c-bd35-af82aa79f546.json:4]: **Nit**: I recommend including a link to the breaking discussion post.

**Review by ianbotsf — APPROVED** [review]

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/engine/internal/HttpClientMetrics.kt [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        "The amount of time requests take to acquire a connection from the pool",
+    )
+
+    /**
+     * The amount of time a request spent queued waiting to be executed by the HTTP client
+     */
+    public val requestsQueuedDuration: DoubleHistogram = meter.createDoubleHistogram(
+        "smithy.client.http.requests.queued_duration",
+        "s",
+        "The amount of time a requests spent queued waiting to be executed by the HTTP client",
```
Comment (lauzadis) [inline runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/engine/internal/HttpClientMetrics.kt:67]: nit: "a request"

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/OperationTelemetryInterceptor.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        val serializeDuration = serializeStart?.elapsedNow() ?: return
+        metrics.serializationDuration.recordSeconds(serializeDuration, perRpcAttributes, metrics.provider.contextManager.current())
+    }
+
+    override fun readAfterAttempt(context: ResponseInterceptorContext<Any, Any, HttpRequest, HttpResponse?>) {
+        metrics.rpcAttempts.add(1L, perRpcAttributes, metrics.provider.contextManager.current())
+        attempts++
+        if (attempts > 1) {
+            metrics.rpcRetryCount.add(1L, perRpcAttributes, metrics.provider.contextManager.current())
+        }
```
Comment (lauzadis) [inline runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/OperationTelemetryInterceptor.kt:96]: I think this is counting the first attempt as a retry, is that correct?

**Review by lauzadis — APPROVED** [review]

**Pull request merged**